RavenDB-18103 refactoring ShardedIndexesContext

ppekrol opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18103

ppekrol committed:
RavenDB-18103 started working on AbstractIndexCreateProcessor

ppekrol committed:
RavenDB-18103 refactored ShardedIndexesContext

ppekrol committed:
RavenDB-18103 further simplification of ShardedIndexesContext

ppekrol committed:
RavenDB-18103 introduced processors

ppekrol committed:
RavenDB-18103 fixing timeseries & counters

ppekrol committed:
RavenDB-18103 cleanup

ppekrol committed:
RavenDB-18103 implemented ValidateStaticIndexAsync

ppekrol committed:
RavenDB-18103 cleanup

arekpalinski commented on src/Raven.Server/Documents/Indexes/IndexContext.cs:
```diff
@@ -0,0 +1,118 @@
+            return IndexType.MapReduce;
+
+        if (definition is MapIndexDefinition)
+            return IndexType.Map;
```
`JavaScriptMap` and  `JavaScriptMapReduce` are missing I think

arekpalinski commented on src/Raven.Server/Documents/Indexes/IndexContext.cs:
```diff
@@ -0,0 +1,118 @@
+    public readonly AbstractStaticIndexBase Compiled;
+}
+
+public class IndexContext
```
My proposals for the name to avoid having one more "context"

- `IndexBasicInfo` 
- `IndexInfoContainer` 
- `IndexInfoHolder`
- `IndexComponents`
- `IndexIngredients`

arekpalinski commented on src/Raven.Server/Documents/Indexes/AbstractIndexCreateProcessor.cs:
```diff
@@ -0,0 +1,187 @@
+
+namespace Raven.Server.Documents.Indexes;
+
+public abstract class AbstractIndexCreateProcessor
```
My proposals for different name:

- `AbstractIndexCreation` / `DatabaseIndexCreation` / `ShardIndexCreation`
- `AbstractIndexCreateDelegate` / `DatabaseIndexCreateDelegate` / `ShardIndexCreateDelegate`
- `AbstractIndexCreationMethods` / `DatabaseIndexCreationMethods` / `ShardIndexCreationMethods`

ppekrol committed:
RavenDB-18103
- renamed *Processor => *Controller
- renamed IndexContext to IndexInformationHolder
- fixed concurrent context usage in WaitForExecutionOfRaftCommandsAsync

ppekrol committed:
RavenDB-18103 use ShardedIndexCreateController

ppekrol committed:
RavenDB-18103 simple mechanism that bumps generation of ScriptRunnerCache

ppekrol committed:
RavenDB-18103 implemented /databases/*/indexes/source

ppekrol committed:
RavenDB-18103 fixing namespace

ppekrol committed:
RavenDB-18103 implemented /databases/*/indexes/performance

ppekrol committed:
RavenDB-18103 do not change url in GetIndexPerformanceStatisticsCommand when index names is empty

gregolsky commented:
test this please

ppekrol committed:
RavenDB-18103 properly wait for the commands to finish

ppekrol committed:
RavenDB-18103 better message in assert
